ffi/derive/src/convert.rs

```diff
@@ -20,140 +27,182 @@ pub fn derive_ffi_type(mut input: DeriveInput) -> TokenStream {
         Data::Enum(item) => {
             let repr = find_attr(&input.attrs, "repr");
 
-            if item.variants.is_empty() {
-                abort!(name, "uninhabited enum's cannot be instantiated");
-            }
-
             if is_fieldless_enum(&item) {
+                if item.variants.len() == 1 {
+                    // NOTE: one-variant fieldless enums have representation of ()
+                    return derive_ffi_type_for_opaque_item(name, &input.generics);
+                }
+                if without_repr(&input.attrs) {
+                    abort!(name, "Fieldless enum missing #[repr(int)]");
+                }
                 if item.variants.iter().any(|v| v.discriminant.is_some()) {
                     abort!(
                         name,
-                        "fieldless enums with explicit discriminants are prohibited"
+                        "Fieldless enums with explicit discriminants are prohibited"
                     );
                 }
+
                 derive_ffi_type_for_fieldless_enum(&input.ident, &item, &repr)
             } else {
                 let local = !is_non_local(&input.attrs);
                 derive_ffi_type_for_data_carrying_enum(&input.ident, input.generics, &item, local)
             }
         }
-        Data::Struct(_) => {
-            if is_owning(&input.attrs, &input.data) {
-                abort!(
-                    name,
-                    "Structure contains raw pointers. If the pointers don't own the data, attach `#[ffi_type(unsafe {non_owning})`"
-                )
-            }
+        Data::Struct(ref item) => {
+            let ffi_type_impl = derive_ffi_type_for_repr_c(&input);
+
+            let repr_c_impl = {
+                let predicates = &mut input.generics.make_where_clause().predicates;
+                let add_bound = |ty| predicates.push(parse_quote! {#ty: iroha_ffi::ReprC});
+
+                match &item.fields {
+                    syn::Fields::Named(fields) => fields
+                        .named
+                        .iter()
+                        .map(|field| &field.ty)
+                        .for_each(add_bound),
+                    syn::Fields::Unnamed(fields) => fields
+                        .unnamed
+                        .iter()
+                        .map(|field| &field.ty)
+                        .for_each(add_bound),
+                    syn::Fields::Unit => {}
```

Comment:
```suggestion
                    syn::Fields::Unit => unreachable!("There can't be any unit-fields in struct")
```

I think it shows better your idea, if I understood it. Without it a reader (me) starts thinking, why do we ingore it